### ISSUE TITLE
InputStream implementation?

### ISSUE THREAD

**Issue by tomas-pluskal** [ms-numpress/ms-numpress#11]
Hi, 
Thanks for providing this library! I wonder if it would be possible to make the Java class implement java.io.FilterInputStream, similar to java.util.zip.InflaterInputStream. It would be quite helpful to perform numpress decompression on-the-fly via streams.
Tomas

**Comment by hroest** [comment]
Hi Tomas, 

curently, ms-numpress is in maintenance mode and there is little active development at the current time. However, we would be more than happy to take pull requests in case you decide to implement something like that - so far I dont think this issue has been critical as in all mass spec data the individual spectrum is a separate entity and there was not really a need to do decompression on the fly as the individual data entity is pretty small and fits very well into memory. Do you have a specific use-case where you think it would make a lot of sense to have on-the-fly decompression?

**Comment by tomas-pluskal** [comment]
Yes, we are working on a mzML access library that uses memory-mapping for fast access. From the memory-mapped file, we use InputStreams that perform Base64 decoding and decompression, so we avoid creating any unnecessary arrays. This way there is little overhead for the garbage collector and the whole data can be streamed without allocating any memory - important when working with multiple large files.

**Comment by hroest** [comment]
That sounds interesting, so this is indeed missing in the current implementation. There are a few caveats as the bytes at the beginning are special and need to be read first as they contain variables used for the encoding and you will therefore have state variables. I dont know enough java to implement something like this myself but it sounds interesting. However, another thought: if you really want to have fast access to individual spectra, you will suffer from base64 conversion and de-compression. Have you looked at this recent paper comparing a few approaches: http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0125108 ? Maybe caching the data or writing it into a sqlite database would give you faster access?

**Comment by tomas-pluskal** [comment]
Caching is of course an option, but our goal is to have a reasonably fast basic access option that does not require additional memory or disk space. Caching can be used as an additional layer for speeding up repeated access.

**Comment by hroest** [comment]
yes, that makes a lot of sense. Maybe @fickludd could weigh in how hard it would be to implement this, to me it seems that it would be terribly difficult, you would just need to implement a class that inherits from https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html and then deliver one byte at a time, correct?

**Comment by tomas-pluskal** [comment]
I guess the most straightforward way would be to copy&paste the InflaterInputStream code: 
http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/zip/InflaterInputStream.java

**Comment by tomas-pluskal** [comment]
In the code, they call the Inflater class to perform the actual decompression. That would have to be replaced by your own code for Numpress decompressing.

**Comment by fickludd** [comment]
@tomas-pluskal This is an interesting idea. I did play around back then with storing spectra as numpressed arrays in memory and only decompressing on access, but in my experiments speed was essentially unchanged. I guess cache locality and decompression cost just about evened out. 

Anyhow, I agree that this should not be too hard to implement by modifying the `InflaterInputStream`. And if you want it to happen, the best bet is to do it yourself, make a PR, and I will happily review and merge it once it seems robust for general public use. 

Another note, If you want fast access I would suggest to have your own binary in-memory format rather than the pure mzML. As @hroest says: you'll pay for Base64, both in decompression and also 33% more memory use. You can still numpress the data =).
